consensus/dolev_strong.go

```diff
@@ -0,0 +1,310 @@
+package consensus
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/consensus/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	//"reflect"
+	"bytes"
+	dsCfg "github.com/spacemeshos/go-spacemesh/consensus/config"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"time"
+	"errors"
+	"fmt"
+)
+
+// DolevStrongConsensus is an implementation of a byzanteen agreement protocol
+type DolevStrongConsensus struct {
+	activeInstances map[string]CAInstance
+	ingressChannel  chan OpaqueMessage
+	net             NetworkConnection
+	dsConfig        dsCfg.Config
+	privateKey      crypto.PrivateKey
+	publicKey       crypto.PublicKey
+	timer           Timer
+	startTime       time.Time
+	abortInstance   chan struct{}
+	abortListening  chan struct{}
+}
+
+// DolevStrongInstance is a struct holding state for a single instance of a dolev strong agreement protocol as a receiver
+type DolevStrongInstance struct {
+	value                   []byte
+	ds                      *DolevStrongConsensus
+	abortProcessingInstance bool
+}
+
+// StartInstance starts an agreement protocol instance
+func (impl *DolevStrongConsensus) StartInstance(msg OpaqueMessage) []byte {
+	impl.startTime = time.Now()
```

Comment:
you should assert that there isn't already an active instance. Also, I think that it should say that by instance you mean sender instance